File: ios/NoctuaSDK/NoctuaSDK/NoctuaService.swift
//
//  NoctuaService.swift
//  NoctuaSDK
//
//  Created by SDK Dev on 31/07/24.
//

import Foundation

struct NoctuaServiceConfig : Decodable {
    let trackerURL: String?
}

class NoctuaService {
    let trackerURL: String
    
    init(config: NoctuaServiceConfig) {
        trackerURL = if config.trackerURL == nil || config.trackerURL!.isEmpty {
            "https://kafka-proxy-poc.noctuaprojects.com/api/v1/events"
        } else {
            config.trackerURL!
        }
    }
    
    func trackAdRevenue(source: String, revenue: Double, currency: String, extraPayload: [String:Encodable]) {
        var payload = extraPayload
        payload["source"] = source
        payload["revenue"] = revenue
        payload["currency"] = currency
        payload["event_name"] = "AdRevenue"
        
        sendEvent(payload: payload)
    }
    
    func trackPurchase(orderId: String, amount: Double, currency: String, extraPayload: [String:Encodable]) {
        var payload = extraPayload
        payload["orderId"] = orderId
        payload["amount"] = amount
        payload["currency"] = currency
        payload["event_name"] = "Purchase"
        
        sendEvent(payload: payload)
    }
    
    func trackCustomEvent(_ eventName: String, payload: [String:Encodable]) {
        var payload = payload
        payload["event_name"] = eventName

        sendEvent(payload: payload)
    }
    
    private func sendEvent(payload: [String:Encodable]) {
        let encoder = JSONEncoder()
        let bodyData = encodeDictionaryToJSON(dictionary: payload)
        
        
        var request = URLRequest(url: URL(string: trackerURL)!)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = bodyData

        let task = URLSession.shared.dataTask(with: request) {(data, response, error) in
            guard let data = data else { return }
        }

        task.resume()
    }
    
    private struct AnyEncodable: Encodable {
        let value: Encodable

        init(_ value: Encodable) {
            self.value = value
        }

        func encode(to encoder: Encoder) throws {
            var container = encoder.singleValueContainer()
            try container.encode(value)
        }
    }

    // Function to encode a dictionary of type [String: Encodable] to JSON
    private func encodeDictionaryToJSON(dictionary: [String: Encodable]) -> Data? {
        let encodableDictionary = dictionary.mapValues { AnyEncodable($0) }
        do {
            let jsonData = try JSONEncoder().encode(encodableDictionary)
            return jsonData
        } catch {
            print("Error encoding to JSON: \(error)")
            return nil
        }
    }
}

File: android/sdk/src/main/java/com/noctuagames/sdk/NoctuaService.kt
package com.noctuagames.sdk

import android.content.Context
import android.util.Log
import com.adjust.sdk.Adjust
import com.google.gson.Gson
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody
import okhttp3.RequestBody.Companion.toRequestBody

data class NoctuaServiceConfig(
    val trackerURL: String,
)

internal class NoctuaService(
    config: NoctuaServiceConfig?,
    private val context: Context,
    private val withAdjust: Boolean = false
) {
    private val trackerURL =
        config?.trackerURL ?: "https://kafka-proxy-poc.noctuaprojects.com/api/v1/events"

    private val deviceInfo = DeviceInfo(context).getDeviceInfoMap()

    fun trackFirstInstall(extraPayload: MutableMap<String, Any> = mutableMapOf()) {
        val sharedPreferences = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        val isFirstInstall = sharedPreferences.getBoolean(KEY_FIRST_INSTALL, true)
        if (isFirstInstall) {
            Log.d(TAG, "This is the first install.")

            trackCustomEvent("Install", extraPayload.toMutableMap())

            // Mark as not first install
            sharedPreferences.edit().putBoolean(KEY_FIRST_INSTALL, false).apply()
        } else {
            Log.d(TAG, "This is not the first install.")
        }
    }

    fun trackAdRevenue(
        source: String,
        revenue: Double,
        currency: String,
        extraPayload: MutableMap<String, Any> = mutableMapOf()
    ) {
        extraPayload["source"] = source
        extraPayload["revenue"] = revenue
        extraPayload["currency"] = currency

        sendEvent("AdRevenue", extraPayload)
    }

    fun trackPurchase(
        orderId: String,
        amount: Double,
        currency: String,
        extraPayload: MutableMap<String, Any> = mutableMapOf()
    ) {
        extraPayload["orderId"] = orderId
        extraPayload["amount"] = amount
        extraPayload["currency"] = currency

        sendEvent("Purchase", extraPayload)
    }

    fun trackCustomEvent(eventName: String, payload: MutableMap<String, Any> = mutableMapOf()) {
        sendEvent(eventName, payload)
    }

    private fun sendEvent(eventName: String, payload: MutableMap<String, Any>) {
        payload["event_name"] = eventName
        payload.putAll(deviceInfo)

        if (withAdjust) {
            payload.putAllAdjustMetadata()
        }

        Log.d(TAG, "Sending event: $payload")

        CoroutineScope(Dispatchers.IO).launch {
            val client = OkHttpClient()
            val json = Gson().toJson(payload)

            val mediaType = "application/json; charset=utf-8".toMediaTypeOrNull()
            val body: RequestBody = json.toRequestBody(mediaType)

            val request = Request.Builder()
                .url(trackerURL)
                .post(body)
                .build()

            try {
                val response = client.newCall(request).execute()
                withContext(Dispatchers.Main) {
                    if (response.isSuccessful) {
                        Log.d(TAG, "Event sent successfully: ${response.body?.string()}")
                    } else {
                        Log.w(TAG, "Failed to send event: ${response.body?.string()}")
                    }
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    private companion object {
        private val TAG = NoctuaService::class.simpleName
        private const val PREFS_NAME = "NoctuaTrackerPrefs"
        private const val KEY_FIRST_INSTALL = "isFirstInstall"
    }
}

private fun MutableMap<String, Any>.putAllAdjustMetadata() {
    Adjust.getAttribution()?.let {
        this["attribution_adid"] = it.adid
        this["attribution_tracker_name"] = it.trackerName
        this["attribution_tracker_token"] = it.trackerToken
        this["attribution_adgroup"] = it.adgroup
        this["attribution_network"] = it.network
        this["attribution_campaign"] = it.campaign
        this["attribution_clicklabel"] = it.clickLabel
        this["attribution_cost_currency"] = it.costCurrency
        this["attribution_cost_type"] = it.costType
        this["attribution_creative"] = it.creative
    }

    this["adid"] = Adjust.getAdid()
}